File: app/src/main/java/com/ltst/instagramgallerysample/MainActivity.java
package com.ltst.instagramgallerysample;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.GridLayoutManager;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

import com.cosic.instagallery.data.GalleryData;
import com.cosic.instagallery.gallery.GalleryAppBarLayout;
import com.cosic.instagallery.gallery.GalleryRecyclerView;
import com.cosic.instagallery.utils.EndlessRecyclerScrollListener;
import com.cosic.instagallery.utils.GridSpacingItemDecoration;
import com.cosic.instagallery.utils.Logger;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        final int spanCount = getResources().getInteger(R.integer.span_count);

        final GalleryAppBarLayout appBarLayout = findViewById(R.id.appbar);
        appBarLayout.setOnCollapseChangeStateListener(new GalleryAppBarLayout.OnCollapseChangeStateListener() {
            @Override
            public void onCollapsed() {
                Logger.d("OnCollapseChangeStateListener: collapsed");
            }

            @Override
            public void onExpended() {
                Logger.d("OnCollapseChangeStateListener: expended");
            }
        });

        final Button collapseButton = findViewById(R.id.collapse_button);
        final Button expandButton = findViewById(R.id.expand_button);
        final GalleryRecyclerView recyclerView = findViewById(R.id.gallery);
        final GridLayoutManager layoutManager = new GridLayoutManager(this, spanCount);
        recyclerView.setLayoutManager(layoutManager);
        recyclerView.addItemDecoration(new GridSpacingItemDecoration(
                spanCount, getResources().getDimensionPixelSize(R.dimen.divider_size), false));

        recyclerView.addOnScrollListener(new EndlessRecyclerScrollListener(layoutManager, 0) {
            @Override
            protected void onScrolledToEndChanged(final boolean isEnded) {
                if (isEnded) {
                    appBarLayout.expand();
                }
            }
        });

        collapseButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                appBarLayout.collapse();
            }
        });

        expandButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                appBarLayout.expand();
            }
        });

        GalleryAdapter adapter = new GalleryAdapter(this);
        recyclerView.setAdapter(adapter);
        adapter.setOnGalleryClickListener(new GalleryAdapter.OnGalleryClickListener() {
            @Override
            public void onClick(final int position, final GalleryData item) { // TODO move to library;
                Toast.makeText(MainActivity.this, "Click by " + item.value, Toast.LENGTH_SHORT).show();
                appBarLayout.expand();
                recyclerView.scrollToPosition(position);
            }
        });

        mockData(adapter);
    }

    private void mockData(GalleryAdapter adapter) {
        List<GalleryData> items = new ArrayList<>();
        int count = 100;
        for (int i = 0; i < count; i++) {
            items.add(new GalleryData(i + 1));
        }
        adapter.addAll(items);
        adapter.notifyDataSetChanged();
    }
}

File: app/src/main/java/com/cosic/instagramgallerysample/MainActivity.kt
package com.cosic.instagramgallerysample

import android.os.Bundle
import android.widget.Button
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.GridLayoutManager
import com.cosic.instagallery.GalleryAppBarLayout
import com.cosic.instagallery.GalleryRecyclerView
import com.cosic.instagallery.EndlessRecyclerScrollListener
import com.cosic.instagallery.Logger
import java.util.ArrayList

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val spanCount = resources.getInteger(R.integer.span_count)

        val appBarLayout = findViewById<GalleryAppBarLayout>(R.id.appbar)
        appBarLayout.setOnCollapseChangeStateListener(object : GalleryAppBarLayout.OnCollapseChangeStateListener {
            override fun onCollapsed() {
                Logger.d("OnCollapseChangeStateListener: collapsed")
            }

            override fun onExpended() {
                Logger.d("OnCollapseChangeStateListener: expended")
            }
        })

        val collapseButton = findViewById<Button>(R.id.collapse_button)
        val expandButton = findViewById<Button>(R.id.expand_button)
        val recyclerView = findViewById<GalleryRecyclerView>(R.id.gallery)
        val layoutManager = GridLayoutManager(this, spanCount)
        recyclerView.layoutManager = layoutManager
        recyclerView.addItemDecoration(GridSpacingItemDecoration(
            spanCount, resources.getDimensionPixelSize(R.dimen.divider_size), false))

        recyclerView.addOnScrollListener(object : EndlessRecyclerScrollListener(layoutManager, 0) {
            override fun onScrolledToEndChanged(isEnded: Boolean) {
                if (isEnded) {
                    appBarLayout.expand()
                }
            }
        })

        collapseButton.setOnClickListener { appBarLayout.collapse() }

        expandButton.setOnClickListener { appBarLayout.expand() }

        val adapter = GalleryAdapter(this)
        recyclerView.adapter = adapter
        adapter.setOnGalleryClickListener(object : GalleryAdapter.OnGalleryClickListener {
            override fun onClick(position: Int, item: GalleryData) {
                // TODO move to library;
                Toast.makeText(this@MainActivity, "Click by " + item.value, Toast.LENGTH_SHORT).show()
                appBarLayout.expand()
                recyclerView.scrollToPosition(position)
            }
        })

        mockData(adapter)
    }

    private fun mockData(adapter: GalleryAdapter) {
        val items = ArrayList<GalleryData>()
        val count = 100
        for (i in 0 until count) {
            items.add(GalleryData(i + 1))
        }
        adapter.addAll(items)
        adapter.notifyDataSetChanged()
    }
}
